libclamav/pdf.c

```diff
@@ -2879,184 +2903,325 @@ static void compute_hash_r6(const char *password, size_t pwlen, const unsigned c
-                }
+            size_t pwlen    = 0;
+            char password[] = "";
 
-                aes_256cbc_decrypt((const unsigned char *)UE, &UE_len, (unsigned char *)(pdf->key), (char *)result2, 32, 0);
-                dbg_printhex("check_user_password: Candidate encryption key", pdf->key, pdf->keylen);
-
-                password_empty = 1;
+            if (NULL == OE) {
+                cli_dbgmsg("check_owner_password: Missing OE value!\n");
+                noisy_warnmsg("check_owner_password: Missing OE value!\n");
+                goto done;
             }
-        }
-    } else if (R == 6) {
-        unsigned char hash[32], validationkey[32];
 
-        size_t pwlen    = 0;
-        char password[] = "";
+            dbg_printhex("OE: ", OE, OE_len);
 
-        if (NULL == UE) {
-            cli_dbgmsg("check_user_password: Missing UE value!\n");
-            noisy_warnmsg("check_user_password: Missing UE value!\n");
-            return;
-        }
-
-        compute_hash_r6(password, pwlen, (const unsigned char *)(U + 32), validationkey);
-        if (!memcmp(U, validationkey, sizeof(validationkey))) {
-
-            compute_hash_r6(password, pwlen, (const unsigned char *)(U + 40), hash);
+            /*
+             * Test the password against the owner key by computing the SHA-256 hash of the UTF-8 password concatenated
+             * with the 8 bytes of owner validation salt, concatenated with the 48-byte U string.
+             */
+            compute_hash_r6(
+                password,
+                pwlen,
+                (const unsigned char *)(O + 32), // owner validation salt
+                validationkey,
+                U);
+
+            /* If the 32-byte result matches the first 32 bytes of the O string, this is the owner password. */
+            if (0 != memcmp(O, validationkey, sizeof(validationkey))) {
+                cli_dbgmsg("check_owner_password: Owner password check did not match!\n");
+                break;
+            }
 
-            if (UE_len != 32) {
-                cli_dbgmsg("check_user_password: UE length is not 32: %zu\n", UE_len);
-                noisy_warnmsg("check_user_password: UE length is not 32: %zu\n", UE_len);
+            /*
+             * Compute an intermediate owner key by computing the SHA-256 hash of the UTF-8 password concatenated with
+             * the 8 bytes of owner key salt, concatenated with the 48-byte U string.
+             */
+            compute_hash_r6(
+                password,
+                pwlen,
+                (const unsigned char *)(O + 40), // owner key salt
+                hash,
+                U);
+
+            if (OE_len != 32) {
+                cli_dbgmsg("check_owner_password: OE length is not 32: %zu\n", OE_len);
+                noisy_warnmsg("check_owner_password: OE length is not 32: %zu\n", OE_len);
             } else {
                 pdf->keylen = 32;
                 pdf->key    = cli_malloc(pdf->keylen);
```

Comment:
probably don't need cli_malloc here, since keylen is fixed.

Reply:
I forgot to reply to this one.  There are 3 other locations where the pdf->key is allocated.  For one of those,  the size is not 32.  I confirmed that for that one it's always less than 32, so... we could probably make it a part of the pdf buffer and not malloc it here.  But I'm okay with keeping it as-is for now.